### ISSUE TITLE
Suboptimal codegen: Unnecessary stack save of non-callee-preserved register

### ISSUE THREAD

**Issue by bdonlan** [rust-lang/rust#73701]
When looking at the generated assembly of the following code (in a release build), I noticed some suboptimal code generation.

```rust
pub fn g(n: Option<u64>) -> u64 {
    n.unwrap()
}
```

This produces the following code (on nightly):

```asm
playground::g:  # @playground::g
# %bb.0:
	pushq	%rax
	testq	%rdi, %rdi
	je	.LBB1_1
# %bb.2:
	movq	%rsi, %rax
	popq	%rcx
	retq

.LBB1_1:
	leaq	.L__unnamed_1(%rip), %rdi
	leaq	.L__unnamed_3(%rip), %rdx
	movl	$43, %esi
	callq	*core::panicking::panic@GOTPCREL(%rip)
	ud2
                                        # -- End function
```

As you can see, the %rax register is saved in the function preamble, then restored to the rcx register. Both rax and rcx registers can be safely clobbered by callees in the SystemV x86_64 calling convention - and in particular, the rax register holds the return value, so it shouldn't have anything interesting for the unwrap panic handler - so it seems strange that these push/pops would be emitted in the first place.

Note that this is not emitted with `unwrap_or`:

```rust
pub fn l(n: Option<u64>) -> u64 {
    n.unwrap_or(42)
}
```

```asm
playground::l:  # @playground::l
# %bb.0:
	testq	%rdi, %rdi
	movl	$42, %eax
	cmovneq	%rsi, %rax
	retq
                                        # -- End function
```

### Meta

I tested this on the rust playground using both stable and nightly in release mode: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=138efc3ed37cca50141741c3e746e332

**Comment by nagisa** [comment]
On x86_64 stack needs to be aligned to 16 bytes when calling functions, so a push (or some other stack aligning instruction) needs to happen _somewhere_, but you're right that it does not need to happen where it currently does. Initially in LLVM codegen the code looks as like this:

```
bb.0 (%ir-block.0):
  successors: %bb.1(0x00000001), %bb.2(0x7fffffff); %bb.1(0.00%), %bb.2(100.00%)
  liveins: $rdi, $rsi
  %1:gr64 = COPY $rsi
  %0:gr64 = COPY $rdi
  TEST64rr %0:gr64, %0:gr64, implicit-def $eflags
  JCC_1 %bb.2, 5, implicit $eflags
  JMP_1 %bb.1


bb.1.bb2.i:
; predecessors: %bb.0
  ADJCALLSTACKDOWN64 0, 0, 0, implicit-def dead $rsp, implicit-def dead $eflags, implicit-def dead $ssp, implicit $rsp, implicit $ssp
  %2:gr64 = MOV32ri64 43
  %3:gr64 = IMPLICIT_DEF
  $rdi = COPY %3:gr64
  $rsi = COPY %2:gr64
  %4:gr64 = IMPLICIT_DEF
  $rdx = COPY %4:gr64
  CALL64m $rip, 1, $noreg, target-flags(x86-gotpcrel) @_ZN4core9panicking5panic17h68e56c2eeba99c8cE, $noreg, <regmask $bh $bl $bp $bph $bpl $bx $ebp $ebx $hbp $hbx $rbp $rbx $r12 $r13 $r14 $r15 $r12b $r13b $r14b $r15b $r12bh $r13bh $r14bh $r15bh $r12d $r13d $r14d $r15d $r12w $r13w $r14w $r15w $r12wh and 3 more...>, implicit $rsp, implicit $ssp, implicit $rdi, implicit $rsi, implicit $rdx, implicit-def $rsp, implicit-def $ssp :: (load 8 from got)
  ADJCALLSTACKUP64 0, 0, implicit-def dead $rsp, implicit-def dead $eflags, implicit-def dead $ssp, implicit $rsp, implicit $ssp


bb.2._ZN4core6option15Option$LT$T$GT$6unwrap17h2ccd910ff0fb2487E.exit:
; predecessors: %bb.0
  $rax = COPY %1:gr64
  RET 0, $rax
```

Note how LLVM knows both that bb.1, which contains the panic call has 0% likeliness (due to the cold annotation) and how `ADJCALLSTACKDOWN/UP64` calls are only present in the `bb.1`. In particular `ADJCALLSTACKUP64` is not necessary and I suspect eventually leads to prolog-epilog inserter pass in setting up prologue epilogue in `bb.0` and `bb.2`, that eventually end up as the extraneous pop instruction.

**Comment by bdonlan** [comment]
After some testing on godbolt, this seems to have regressed between rustc 1.26.0 and 1.27.0.

rustc 1.26.0 assembly:

```asm
example::g:
        cmp     qword ptr [rdi], 1
        jne     .LBB0_1
        mov     rax, qword ptr [rdi + 8]
        ret
.LBB0_1:
        push    rbp
        mov     rbp, rsp
        lea     rdi, [rip + .Lbyte_str.2]
        call    core::panicking::panic@PLT
        ud2

.Lbyte_str.0:
        .ascii  "called `Option::unwrap()` on a `None` value"

.Lbyte_str.1:
        .ascii  "libcore/option.rs"

.Lbyte_str.2:
        .quad   .Lbyte_str.0
        .asciz  "+\000\000\000\000\000\000"
        .quad   .Lbyte_str.1
        .asciz  "\021\000\000\000\000\000\000\000O\001\000\000\025\000\000"
```

rustc 1.27.0:
```asm
example::g:
        push    rax
        test    rdi, rdi
        je      .LBB0_1
        mov     rax, rsi
        pop     rcx
        ret
.LBB0_1:
        lea     rdi, [rip + .Lbyte_str.2]
        call    core::panicking::panic@PLT
        ud2

.Lbyte_str.0:
        .ascii  "called `Option::unwrap()` on a `None` value"

.Lbyte_str.1:
        .ascii  "libcore/option.rs"

.Lbyte_str.2:
        .quad   .Lbyte_str.0
        .asciz  "+\000\000\000\000\000\000"
        .quad   .Lbyte_str.1
        .asciz  "\021\000\000\000\000\000\000\000O\001\000\000\025\000\000"
```

**Comment by andjo403** [comment]
this seems to have been fixed from rustc 1.52 see https://godbolt.org/z/verPfvxaM

**Comment by nikic** [comment]
Closing per previous comment -- this doesn't look like something we'd want a codegen test for.
